# Data Science Fundamentals Assessment
**Topics:** DS lifecycle · Python basics · Statistics · Data types · Best practices · Practical assignments

**Reproducibility:** all random data uses a fixed seed; only `numpy`, `pandas` and the standard library are required.

## 1. Data Science Fundamentals

### 1.1 The Data Science Lifecycle
| Stage | Description | Example (customer churn) |
|---|---|---|
| 1. Problem definition | Turn a business question into a measurable goal | Reduce monthly churn by 10% |
| 2. Data collection | Gather data from databases, APIs, files, sensors | CRM records, usage logs |
| 3. Data cleaning | Fix missing values, duplicates, wrong types, outliers | Impute missing tenure |
| 4. Exploratory analysis (EDA) | Summaries, distributions, relationships | Churn rate by plan type |
| 5. Modelling | Build statistical / ML models | Logistic regression |
| 6. Evaluation | Measure performance with suitable metrics | Precision, recall, AUC |
| 7. Deployment & communication | Ship the model, report findings | Dashboard, weekly retention list |
| 8. Monitoring | Track drift and re-train | Alert if accuracy drops |

The lifecycle is **iterative**: findings in EDA often send you back to data collection or cleaning.

### 1.2 Types of Analysis
| Type | Question | Example | Typical techniques |
|---|---|---|---|
| Descriptive | What happened? | Monthly sales totals | Mean, counts, charts |
| Diagnostic | Why did it happen? | Why did sales dip in March? | Drill-down, correlation |
| Predictive | What will happen? | Next-quarter demand | Regression, classification |
| Prescriptive | What should we do? | Optimal price / stock level | Optimisation, simulation |

### 1.3 Industry Use Cases
- **Healthcare:** predicting readmission risk, medical image analysis
- **Finance:** fraud detection, credit scoring
- **Retail / e-commerce:** recommendation engines, demand forecasting
- **Manufacturing:** predictive maintenance from sensor data
- **Transport / logistics:** route optimisation, ETA prediction
- **Telecom / SaaS:** churn prediction, customer segmentation

### 1.4 Role of Data Science in Real-World Problems
Data science converts raw data into **decisions**: it quantifies uncertainty, finds patterns humans miss at scale, automates repetitive judgement, and lets organisations test hypotheses with evidence instead of intuition. It complements, rather than replaces, domain expertise.

## 2. Python Programming Basics

### 2.1 Variables, Data Types and Operators

In [1]:
age = 29                 # int
height_m = 1.72          # float
name = "Asha"            # str
is_student = False       # bool
nothing = None           # NoneType

for v in (age, height_m, name, is_student, nothing):
    print(f"{v!r:>8}  ->  {type(v).__name__}")

# Arithmetic, comparison, logical operators
a, b = 17, 5
print("\nArithmetic :", a + b, a - b, a * b, a / b, a // b, a % b, a ** 2)
print("Comparison :", a > b, a == b, a != b)
print("Logical    :", (a > 10) and (b < 10), (a < 10) or (b < 10), not (a > b))
print("Membership :", 'sh' in name)

      29  ->  int
    1.72  ->  float
  'Asha'  ->  str
   False  ->  bool
    None  ->  NoneType

Arithmetic : 22 12 85 3.4 3 2 289
Comparison : True False True
Logical    : True True False
Membership : True


### 2.2 Conditional Statements

In [2]:
def grade(score):
    """Return a letter grade for a numeric score (0-100)."""
    if score >= 90:
        return "A"
    elif score >= 75:
        return "B"
    elif score >= 60:
        return "C"
    return "F"

print([(s, grade(s)) for s in (95, 80, 62, 40)])

[(95, 'A'), (80, 'B'), (62, 'C'), (40, 'F')]


### 2.3 Loops

In [3]:
# for loop with range
squares = [n ** 2 for n in range(1, 6)]      # list comprehension
print("Squares:", squares)

# while loop: how many halvings until value < 1?
value, steps = 100, 0
while value >= 1:
    value /= 2
    steps += 1
print("Halvings needed:", steps)

# break / continue
for n in range(1, 10):
    if n % 2 == 0:
        continue
    if n > 7:
        break
    print("odd:", n, end="  ")
print()

Squares: [1, 4, 9, 16, 25]
Halvings needed: 7
odd: 1  odd: 3  odd: 5  odd: 7  


### 2.4 Functions

In [4]:
def describe(numbers, round_to=2):
    """Return a dict with the count, mean, min and max of `numbers`.

    Parameters
    ----------
    numbers : list[float]
    round_to : int, number of decimal places for the mean
    """
    if not numbers:
        raise ValueError("numbers must not be empty")
    return {
        "count": len(numbers),
        "mean": round(sum(numbers) / len(numbers), round_to),
        "min": min(numbers),
        "max": max(numbers),
    }

print(describe([4, 8, 15, 16, 23, 42]))

square = lambda x: x ** 2           # lambda (anonymous function)
print(list(map(square, [1, 2, 3])))

{'count': 6, 'mean': 18.0, 'min': 4, 'max': 42}
[1, 4, 9]


### 2.5 Lists, Tuples, Dictionaries and Sets

In [5]:
# List - ordered, mutable
scores = [88, 92, 79]
scores.append(95); scores.sort(reverse=True)
print("List      :", scores, "| top-2:", scores[:2])

# Tuple - ordered, immutable (good for fixed records)
point = (28.6, 77.2)
lat, lon = point                      # unpacking
print("Tuple     :", point, "-> lat =", lat)

# Dictionary - key/value lookup
student = {"name": "Asha", "age": 29, "courses": ["DS", "Stats"]}
student["city"] = "Pune"
print("Dict      :", student, "| age:", student.get("age"))
for k, v in student.items():
    pass

# Set - unique, unordered; fast membership & set algebra
a, b = {1, 2, 3, 4}, {3, 4, 5}
print("Set       : union", a | b, "| intersection", a & b, "| diff", a - b)
print("Dedupe    :", list(set([1, 1, 2, 2, 3])))

List      : [95, 92, 88, 79] | top-2: [95, 92]
Tuple     : (28.6, 77.2) -> lat = 28.6
Dict      : {'name': 'Asha', 'age': 29, 'courses': ['DS', 'Stats'], 'city': 'Pune'} | age: 29
Set       : union {1, 2, 3, 4, 5} | intersection {3, 4} | diff {1, 2}
Dedupe    : [1, 2, 3]


### 2.6 Basic File Handling

In [6]:
import csv, os, tempfile

path = os.path.join(tempfile.gettempdir(), "students.csv")

# Write
rows = [("name", "score"), ("Asha", 92), ("Ravi", 78), ("Meera", 85)]
with open(path, "w", newline="") as f:          # 'with' closes the file automatically
    csv.writer(f).writerows(rows)

# Read
with open(path, newline="") as f:
    data = list(csv.DictReader(f))
print(data)

# Error handling when a file is missing
try:
    open("does_not_exist.txt")
except FileNotFoundError as err:
    print("Handled:", type(err).__name__)

[{'name': 'Asha', 'score': '92'}, {'name': 'Ravi', 'score': '78'}, {'name': 'Meera', 'score': '85'}]
Handled: FileNotFoundError


## 3. Statistical Foundations
We use a small example: exam scores for 15 students (one deliberate outlier).

In [7]:
import numpy as np
import pandas as pd
from statistics import mean, median, mode, pvariance, variance, stdev

scores = [55, 62, 65, 68, 70, 70, 72, 75, 78, 80, 82, 85, 88, 90, 150]  # 150 = outlier

### 3.1 Central Tendency and Spread

In [8]:
print(f"Mean              : {mean(scores):.2f}")
print(f"Median            : {median(scores):.2f}")
print(f"Mode              : {mode(scores)}")
print(f"Range             : {max(scores) - min(scores)}")
print(f"Variance (sample) : {variance(scores):.2f}")
print(f"Std dev (sample)  : {stdev(scores):.2f}")
print(f"Variance (pop.)   : {pvariance(scores):.2f}")

# Manual check of the sample variance formula  s^2 = sum((x - mean)^2) / (n - 1)
m = mean(scores)
manual = sum((x - m) ** 2 for x in scores) / (len(scores) - 1)
print("Manual variance matches:", np.isclose(manual, variance(scores)))

Mean              : 79.33
Median            : 75.00
Mode              : 70
Range             : 95
Variance (sample) : 478.38
Std dev (sample)  : 21.87
Variance (pop.)   : 446.49
Manual variance matches: True


**Interpretation:** the mean (≈ 79.3) is pulled above the median (75) by the outlier at 150, so the **median is the more robust** measure here. Use the *sample* formula (n−1) when data is a sample of a larger population.

### 3.2 Percentiles, Quartiles and Outlier Detection (IQR rule)

In [9]:
q1, q2, q3 = np.percentile(scores, [25, 50, 75])
iqr = q3 - q1
lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
print(f"Q1={q1}, Median={q2}, Q3={q3}, IQR={iqr}")
print(f"Outlier fences: [{lower}, {upper}]")
print("Outliers:", [x for x in scores if x < lower or x > upper])
print("90th percentile:", np.percentile(scores, 90))

Q1=69.0, Median=75.0, Q3=83.5, IQR=14.5
Outlier fences: [47.25, 105.25]
Outliers: [150]
90th percentile: 89.2


### 3.3 Probability Basics

In [10]:
# Classical probability: P(event) = favourable outcomes / total outcomes
p_even_die = 3 / 6
print("P(even on a fair die)           =", p_even_die)

# Independent events: P(A and B) = P(A) * P(B)
print("P(two heads in two coin flips)  =", 0.5 * 0.5)

# Complement: P(not A) = 1 - P(A)
print("P(at least one 6 in 3 rolls)    =", round(1 - (5 / 6) ** 3, 4))

# Conditional probability: P(A|B) = P(A and B) / P(B)
# 60% of customers use mobile; 30% use mobile AND buy.
p_buy_given_mobile = 0.30 / 0.60
print("P(buy | mobile)                 =", p_buy_given_mobile)

# Simulation check of the law of large numbers
rng = np.random.default_rng(42)
rolls = rng.integers(1, 7, size=100_000)
print("Simulated P(even)               =", round((rolls % 2 == 0).mean(), 4))

P(even on a fair die)           = 0.5
P(two heads in two coin flips)  = 0.25
P(at least one 6 in 3 rolls)    = 0.4213
P(buy | mobile)                 = 0.5
Simulated P(even)               = 0.4996


### 3.4 Correlation

In [11]:
rng = np.random.default_rng(42)
hours = rng.uniform(1, 10, 50)
marks = 40 + 5 * hours + rng.normal(0, 5, 50)       # strong positive relationship
noise = rng.normal(0, 1, 50)                        # unrelated variable

df_corr = pd.DataFrame({"hours_studied": hours, "marks": marks, "noise": noise})
print(df_corr.corr().round(3))

               hours_studied  marks  noise
hours_studied          1.000  0.958  0.054
marks                  0.958  1.000  0.103
noise                  0.054  0.103  1.000


**Interpretation:** Pearson *r* ranges from −1 to +1. `hours_studied` vs `marks` shows a strong positive correlation, while `noise` is close to 0. **Correlation ≠ causation**: a high *r* only shows association; confounders or coincidence can produce it.

## 4. Data Types and Structures

| Type | Description | Examples | Typical handling |
|---|---|---|---|
| Numerical – continuous | Measurable, any value in range | height, price | Mean, std, scaling |
| Numerical – discrete | Countable integers | number of orders | Counts, Poisson-type models |
| Categorical (nominal) | Labels with no order | city, colour | One-hot encoding |
| Ordinal | Ordered categories | Low/Med/High | Ordered encoding (preserve rank) |
| Datetime | Timestamps | order date | Parse, extract parts, resample |
| Text | Free-form strings | reviews | Clean, tokenise, vectorise |
| Boolean | True/False | is_active | 0/1 |

In [12]:
df = pd.DataFrame({
    "customer_id": [101, 102, 103, 104, 105, 106],
    "age":         [25, 34, None, 45, 29, 52],                 # numerical
    "city":        ["Pune", "Delhi", "Pune", "Mumbai", "Delhi", "Pune"],   # nominal
    "satisfaction":["Low", "High", "Medium", "High", "Low", "Medium"],     # ordinal
    "signup_date": ["2024-01-15", "2024-02-20", "2024-03-05",
                    "2024-03-28", "2024-04-10", "2024-05-01"],  # datetime (as text)
    "review":      ["Great service!", "Too slow...", "Okay, not bad",
                    "LOVED it", "Poor support", "fine"],
})
print(df.dtypes, "\n")

# Convert to appropriate types
df["city"] = df["city"].astype("category")
df["satisfaction"] = pd.Categorical(df["satisfaction"],
                                    categories=["Low", "Medium", "High"], ordered=True)
df["signup_date"] = pd.to_datetime(df["signup_date"])
print(df.dtypes)

customer_id       int64
age             float64
city                str
satisfaction        str
signup_date         str
review              str
dtype: object 

customer_id              int64
age                    float64
city                  category
satisfaction          category
signup_date     datetime64[us]
review                     str
dtype: object


In [13]:
# Handle each type appropriately
df["satisfaction_code"] = df["satisfaction"].cat.codes            # ordinal -> ranks 0,1,2
city_dummies = pd.get_dummies(df["city"], prefix="city")          # nominal -> one-hot
df["signup_month"] = df["signup_date"].dt.month_name()            # datetime -> feature
df["review_clean"] = (df["review"].str.lower()
                      .str.replace(r"[^a-z ]", "", regex=True).str.strip())  # text cleaning
df["review_word_count"] = df["review_clean"].str.split().str.len()

print(pd.concat([df[["customer_id", "satisfaction", "satisfaction_code",
                     "signup_month", "review_clean", "review_word_count"]],
                 city_dummies], axis=1).to_string(index=False))
print("\nIs 'High' > 'Low'?", df.loc[1, "satisfaction"] > df.loc[0, "satisfaction"])

 customer_id satisfaction  satisfaction_code signup_month  review_clean  review_word_count  city_Delhi  city_Mumbai  city_Pune
         101          Low                  0      January great service                  2       False        False       True
         102         High                  2     February      too slow                  2        True        False      False
         103       Medium                  1        March  okay not bad                  3       False        False       True
         104         High                  2        March      loved it                  2       False         True      False
         105          Low                  0        April  poor support                  2        True        False      False
         106       Medium                  1          May          fine                  1       False        False       True

Is 'High' > 'Low'? False


### Data Structures Overview
- **Python built-ins:** list, tuple, dict, set, for general-purpose work.
- **NumPy `ndarray`:** fast, homogeneous, vectorised numeric arrays.
- **pandas `Series` / `DataFrame`:** labelled 1-D / 2-D tabular data with mixed types, the main workhorse for analysis.

In [14]:
arr = np.array([1, 2, 3, 4])
print("Vectorised:", arr * 2, "| mean:", arr.mean())
s = pd.Series([10, 20, 30], index=["a", "b", "c"])
print(s["b"], "|", type(df).__name__, df.shape)

Vectorised: [2 4 6 8] | mean: 2.5
20 | DataFrame (6, 10)


## 5. Data Science Best Practices

| Practice | Poor | Better |
|---|---|---|
| Meaningful names | `d`, `x1`, `tmp` | `sales_df`, `avg_order_value` |
| Small, documented functions | one 200-line cell | functions with docstrings |
| Don't modify raw data | overwrite source file | work on `.copy()`, save cleaned output separately |
| Reproducibility | random results each run | fixed seeds, pinned library versions, notebook that runs top-to-bottom |
| Constants not magic numbers | `if x > 0.37` | `CHURN_THRESHOLD = 0.37` |
| Validate data | assume it is clean | assertions / checks on shape, ranges, nulls |
| Responsible interpretation | "X causes Y" from a correlation | state limits, sample size, bias, uncertainty |

In [15]:
SEED = 42
MIN_VALID_AGE, MAX_VALID_AGE = 0, 120

def clean_customers(raw_df):
    """Return a cleaned copy of the customer table (raw data is not modified).

    Steps: drop duplicate ids, impute missing age with the median, validate range.
    """
    clean_df = raw_df.copy()
    clean_df = clean_df.drop_duplicates(subset="customer_id")
    median_age = clean_df["age"].median()
    clean_df["age"] = clean_df["age"].fillna(median_age)
    assert clean_df["age"].between(MIN_VALID_AGE, MAX_VALID_AGE).all(), "Age out of range"
    return clean_df

cleaned = clean_customers(df[["customer_id", "age"]])
print("Raw missing ages    :", df["age"].isna().sum())
print("Cleaned missing ages:", cleaned["age"].isna().sum())

Raw missing ages    : 1
Cleaned missing ages: 0


**Responsible interpretation checklist:** Is the sample representative? Could a confounder explain the pattern? Are outliers distorting the summary? Is the result practically (not just statistically) meaningful? Have I reported uncertainty and limitations honestly? Does the data contain personal information that must be protected?

## 6. Fundamentals Assignments

### Assignment 1 – Python programming
**Task:** Write a function that takes a sentence and returns the word frequencies (ignoring case and punctuation), then report the top 3 words.

In [16]:
import string
from collections import Counter

def word_frequencies(text):
    """Return a Counter of lower-cased words with punctuation removed."""
    cleaned = text.lower().translate(str.maketrans("", "", string.punctuation))
    return Counter(cleaned.split())

text = "Data is the new oil. Data without analysis is just data, and analysis needs data!"
print(word_frequencies(text).most_common(3))

[('data', 4), ('is', 2), ('analysis', 2)]


### Assignment 2 – Statistics
**Task:** For daily sales values, compute the summary statistics, detect outliers with the IQR rule, and compare the mean with and without the outliers.

In [17]:
sales = pd.Series([220, 240, 235, 250, 245, 230, 260, 255, 238, 900])  # 900 = promo spike
q1, q3 = sales.quantile([0.25, 0.75])
iqr = q3 - q1
mask = sales.between(q1 - 1.5 * iqr, q3 + 1.5 * iqr)

print(sales.describe().round(2))
print("\nOutliers:", sales[~mask].tolist())
print(f"Mean with outliers   : {sales.mean():.1f}")
print(f"Mean without outliers: {sales[mask].mean():.1f}")

count     10.00
mean     307.30
std      208.59
min      220.00
25%      235.75
50%      242.50
75%      253.75
max      900.00
dtype: float64

Outliers: [900]
Mean with outliers   : 307.3
Mean without outliers: 241.4


**Conclusion:** one promotional spike lifts the mean from about 241 to 307, well above any typical day. Report the median or investigate the spike separately rather than silently deleting it.

### Assignment 3 – Data types and data manipulation
**Task:** Using a small orders dataset: fix types, handle missing values, filter, group, and summarise.

In [18]:
rng = np.random.default_rng(SEED)
orders = pd.DataFrame({
    "order_id": range(1, 13),
    "order_date": pd.date_range("2024-01-01", periods=12, freq="10D").astype(str),
    "region": rng.choice(["North", "South", "East", "West"], 12),
    "category": rng.choice(["Electronics", "Clothing", "Grocery"], 12),
    "amount": rng.integers(200, 2000, 12).astype(float),
    "rating": rng.choice([1, 2, 3, 4, 5, None], 12),
})
orders.loc[2, "amount"] = np.nan          # inject a missing value

# 1. Fix types
orders["order_date"] = pd.to_datetime(orders["order_date"])
orders["region"] = orders["region"].astype("category")

# 2. Missing values
print("Missing per column:\n", orders.isna().sum(), "\n")
orders["amount"] = orders["amount"].fillna(orders["amount"].median())

# 3. Feature engineering + filtering
orders["month"] = orders["order_date"].dt.to_period("M")
high_value = orders[orders["amount"] > 1000]
print("High-value orders:", len(high_value), "of", len(orders), "\n")

# 4. Group and aggregate
summary = (orders.groupby("category", observed=True)["amount"]
           .agg(orders="count", total="sum", average="mean")
           .round(1).sort_values("total", ascending=False))
print(summary, "\n")
print(orders.groupby("month")["amount"].sum().round(0))

Missing per column:
 order_id      0
order_date    0
region        0
category      0
amount        1
rating        3
dtype: int64 

High-value orders: 8 of 12 

             orders   total  average
category                            
Grocery           6  7149.0   1191.5
Clothing          4  3353.0    838.2
Electronics       2  2796.0   1398.0 

month
2024-01    5825.0
2024-02    2179.0
2024-03    3182.0
2024-04    2112.0
Freq: M, Name: amount, dtype: float64


### Assignment 4 – Basic analysis (EDA) with interpretation
**Task:** Explore the relationship between advertising spend and revenue.

In [19]:
rng = np.random.default_rng(SEED)
ad_spend = rng.uniform(10, 100, 60)                       # in thousands
revenue = 50 + 3.2 * ad_spend + rng.normal(0, 25, 60)     # in thousands
eda = pd.DataFrame({"ad_spend": ad_spend, "revenue": revenue})

print(eda.describe().round(2), "\n")
r = eda["ad_spend"].corr(eda["revenue"])
print(f"Pearson r = {r:.3f}")

# Simple linear fit: revenue = slope * ad_spend + intercept
slope, intercept = np.polyfit(eda["ad_spend"], eda["revenue"], 1)
r2 = r ** 2
print(f"Fit: revenue = {slope:.2f} * ad_spend + {intercept:.1f}   (R^2 = {r2:.2f})")

       ad_spend  revenue
count     60.00    60.00
mean      56.78   226.53
std       25.53    83.47
min       10.66    40.94
25%       34.58   160.41
50%       56.36   232.58
75%       79.76   292.24
max       97.81   377.63 

Pearson r = 0.974
Fit: revenue = 3.19 * ad_spend + 45.6   (R^2 = 0.95)


**Interpretation:** there is a strong positive association (*r* ≈ 0.97, R² ≈ 0.95): each extra 1k of ad spend goes with roughly 3k more revenue in this (synthetic) data. This is **association, not proven causation**; seasonality, pricing, or market size could also drive both variables. The sample is small and the linear fit should only be applied within the observed spend range (10k–100k).

---
## Summary of Learning
- Understood the DS lifecycle, four analysis types, and real-world use cases.
- Applied core Python: variables, operators, control flow, functions, collections, file I/O, error handling.
- Computed and interpreted descriptive statistics, percentiles, outliers, probability, and correlation.
- Identified numerical, categorical, ordinal, datetime, and text data and prepared each correctly.
- Followed best practices: clear naming, docstrings, copies of raw data, validation, fixed seeds, cautious interpretation.
- Completed four practical assignments combining all of the above.